### ISSUE TITLE
util/log: files past the configured `max-group-size` are not deleted

### ISSUE THREAD

**Issue by manish-k-sharma** [cockroachdb/cockroach#68259]
**Describe the problem**

Cockroach DB is not cleaning up log files when the groups get big.
We have the `max-group-size` set to 1Gib, but our cockroach DB nodes are getting their log disk filled up with sql audits taking up most of the disk (12GB in one case). 

We suspect that the issue started when we upgraded to v21.1. 

This problem is resulting in our databases going down when the log disk gets filled up.

**To Reproduce**

We launch our cockroach DB nodes with docker-compose

```
version: "3.1"
services:
  cockroachdb:
    restart: always
    image: cockroachdb/cockroach:v21.1.6
    network_mode: host
    volumes:
    - /data/cockroach-data:/cockroach/cockroach-data
    - /data/certs:/cockroach/certs
    - /logs/cockroachdb:/logs/cockroachdb
    ports:
    - 26257
    - 8081
    command:
    - 'start'
    - '--locality=region=<region>,zone=<zone>'
    - '--advertise-addr=<node-ip>'
    - '--http-addr=:8081'
    - '--join=<nodes-list>'
    - '--certs-dir=/cockroach/certs'
    - '--log=file-defaults: {dir: /logs/cockroachdb, max-file-size: 100MiB, max-group-size: 1GiB}'

```
Checking the audit log files disk space
```
sudo find /logs/cockroachdb/ -type f -name 'cockroach-sql-audit*' -exec du -ch {} + | grep total$
```
gives 12GB

**Expected behavior**
We expected the log size to be a maximum of 1GB and that the old log files are automatically removed.

**Environment:**
 - CockroachDB version [e.g. 21.1.x]
 - Server OS: Ubuntu 20.04, cockroach runs inside docker container
 - Client app [e.g. `cockroach sql`, JDBC, ...]

**Additional context**
The databases go down once the log disks get filled up.

**Comment by knz** [comment]
@manish-k-sharma I think it's possible that the SQL audit log config does not inherit `file-defaults`. Can you try setting its own `max-group-size` setting separately?

**Comment by manish-k-sharma** [comment]
@knz will this be the right way to do this?
```
    command:
    - 'start'
    - '--locality=region=<region>,zone=<zone>'
    - '--advertise-addr=<node-ip>'
    - '--http-addr=:8081'
    - '--join=<nodes-list>'
    - '--certs-dir=/cockroach/certs'
    - '--log={file-defaults: {dir: /logs/cockroachdb, max-file-size: 100MiB, max-group-size: 1GiB}, sinks: {file-groups: { sql-audit: {channels: [SENSITIVE_ACCESS], dir: /logs/cockroachdb, max-file-size: 100MiB, max-group-size: 1GiB}}}}' 
```

Logs are still not cleared.

**Comment by knz** [comment]
@maniksurtani can you provide the output of `ls -l` on the audit log directory?

**Comment by manish-k-sharma** [comment]
@knz 
```
lrwxrwxrwx 1 root root        87 Aug  5 03:43 cockroach.log -> cockroach.cockroachdb-db-0.root.2021-08-05T03_43_51Z.000001.log
-rw-r----- 1 root root     91315 Jul 26 02:53 cockroach.cockroachdb-db-0.root.2021-07-26T02_49_44Z.000001.log
-rw-r----- 1 root root 104849989 Aug  4 20:08 cockroach.cockroachdb-db-0.root.2021-07-26T02_53_09Z.000001.log
-rw-r----- 1 root root  43651072 Aug  4 21:54 cockroach.cockroachdb-db-0.root.2021-08-04T20_08_53Z.000001.log
-rw-r----- 1 root root   4747329 Aug  5 14:28 cockroach.cockroachdb-db-0.root.2021-08-05T03_43_51Z.000001.log
lrwxrwxrwx 1 root root        94 Jul 26 02:53 cockroach-pebble.log -> cockroach-pebble.cockroachdb-db-0.root.2021-07-26T02_53_09Z.000001.log
-rw-r----- 1 root root      6091 Jul 26 02:51 cockroach-pebble.cockroachdb-db-0.root.2021-07-26T02_49_46Z.000001.log
-rw-r----- 1 root root   7331360 Aug  4 18:09 cockroach-pebble.cockroachdb-db-0.root.2021-07-26T02_53_09Z.000001.log
lrwxrwxrwx 1 root root        97 Aug  5 12:13 cockroach-sql-audit.log -> cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T12_13_36Z.000001.log
-rw-r----- 1 root root 104857173 Jul 26 05:47 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T02_54_05Z.000001.log
-rw-r----- 1 root root 104857140 Jul 26 08:39 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T05_47_49Z.000001.log
-rw-r----- 1 root root 104857530 Jul 26 11:32 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T08_39_22Z.000001.log
-rw-r----- 1 root root 104857519 Jul 26 14:29 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T11_32_40Z.000001.log
-rw-r----- 1 root root 104857131 Jul 26 17:22 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T14_29_48Z.000001.log
-rw-r----- 1 root root 104857444 Jul 26 20:21 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T17_22_20Z.000001.log
-rw-r----- 1 root root 104857450 Jul 26 23:08 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T20_21_54Z.000001.log
-rw-r----- 1 root root 104857481 Jul 27 02:12 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-26T23_08_12Z.000001.log
-rw-r----- 1 root root 104857117 Jul 27 05:04 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T02_12_44Z.000001.log
-rw-r----- 1 root root 104857476 Jul 27 08:00 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T05_04_33Z.000001.log
-rw-r----- 1 root root 104857099 Jul 27 10:59 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T08_00_30Z.000001.log
-rw-r----- 1 root root 104857138 Jul 27 13:43 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T10_59_38Z.000001.log
-rw-r----- 1 root root 104857107 Jul 27 16:35 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T13_43_46Z.000001.log
-rw-r----- 1 root root 104857315 Jul 27 19:28 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T16_35_21Z.000001.log
-rw-r----- 1 root root 104857529 Jul 27 22:14 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T19_28_04Z.000001.log
-rw-r----- 1 root root 104857502 Jul 28 01:00 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-27T22_14_15Z.000001.log
-rw-r----- 1 root root 104857140 Jul 28 03:51 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T01_00_27Z.000001.log
-rw-r----- 1 root root 104857430 Jul 28 06:40 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T03_51_59Z.000001.log
-rw-r----- 1 root root 104857516 Jul 28 09:34 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T06_40_58Z.000001.log
-rw-r----- 1 root root 104857467 Jul 28 12:34 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T09_34_59Z.000001.log
-rw-r----- 1 root root 104857187 Jul 28 15:27 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T12_34_33Z.000001.log
-rw-r----- 1 root root 104857104 Jul 28 18:24 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T15_27_11Z.000001.log
-rw-r----- 1 root root 104857396 Jul 28 21:18 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T18_24_53Z.000001.log
-rw-r----- 1 root root 104857435 Jul 29 00:14 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-28T21_18_25Z.000001.log
-rw-r----- 1 root root 104857306 Jul 29 03:11 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T00_14_11Z.000001.log
-rw-r----- 1 root root 104857571 Jul 29 05:56 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T03_11_27Z.000001.log
-rw-r----- 1 root root 104857469 Jul 29 08:54 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T05_56_34Z.000001.log
-rw-r----- 1 root root 104857516 Jul 29 11:49 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T08_54_02Z.000001.log
-rw-r----- 1 root root 104857494 Jul 29 14:53 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T11_49_55Z.000001.log
-rw-r----- 1 root root 104857060 Jul 29 17:38 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T14_53_44Z.000001.log
-rw-r----- 1 root root 104857561 Jul 29 20:32 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T17_38_19Z.000001.log
-rw-r----- 1 root root 104857068 Jul 29 23:37 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T20_32_05Z.000001.log
-rw-r----- 1 root root 104857366 Jul 30 02:37 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-29T23_37_43Z.000001.log
-rw-r----- 1 root root 104857566 Jul 30 05:31 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T02_37_05Z.000001.log
-rw-r----- 1 root root 104857094 Jul 30 08:27 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T05_31_52Z.000001.log
-rw-r----- 1 root root 104857383 Jul 30 11:19 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T08_27_12Z.000001.log
-rw-r----- 1 root root 104857487 Jul 30 14:07 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T11_19_48Z.000001.log
-rw-r----- 1 root root 104857479 Jul 30 17:06 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T14_08_00Z.000001.log
-rw-r----- 1 root root 104857139 Jul 30 19:52 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T17_06_53Z.000001.log
-rw-r----- 1 root root 104857359 Jul 30 22:39 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T19_52_32Z.000001.log
-rw-r----- 1 root root 104857163 Jul 31 01:34 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-30T22_39_58Z.000001.log
-rw-r----- 1 root root 104857213 Jul 31 04:19 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T01_34_00Z.000001.log
-rw-r----- 1 root root 104857276 Jul 31 07:08 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T04_19_42Z.000001.log
-rw-r----- 1 root root 104857256 Jul 31 09:58 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T07_08_47Z.000001.log
-rw-r----- 1 root root 104857311 Jul 31 12:42 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T09_58_28Z.000001.log
-rw-r----- 1 root root 104857309 Jul 31 15:35 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T12_42_34Z.000001.log
-rw-r----- 1 root root 104857319 Jul 31 18:31 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T15_35_22Z.000001.log
-rw-r----- 1 root root 104857339 Jul 31 21:27 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T18_31_05Z.000001.log
-rw-r----- 1 root root 104857476 Aug  1 00:18 cockroach-sql-audit.cockroachdb-db-0.root.2021-07-31T21_27_11Z.000001.log
-rw-r----- 1 root root 104857111 Aug  1 03:17 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T00_18_27Z.000001.log
-rw-r----- 1 root root 104857539 Aug  1 06:02 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T03_17_00Z.000001.log
-rw-r----- 1 root root 104857070 Aug  1 08:56 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T06_02_51Z.000001.log
-rw-r----- 1 root root 104857512 Aug  1 11:53 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T08_56_35Z.000001.log
-rw-r----- 1 root root 104857087 Aug  1 14:44 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T11_53_22Z.000001.log
-rw-r----- 1 root root 104857377 Aug  1 17:33 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T14_44_29Z.000001.log
-rw-r----- 1 root root 104857408 Aug  1 20:21 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T17_33_15Z.000001.log
-rw-r----- 1 root root 104857552 Aug  1 23:06 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T20_21_38Z.000001.log
-rw-r----- 1 root root 104857267 Aug  2 01:57 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-01T23_06_47Z.000001.log
-rw-r----- 1 root root 104857290 Aug  2 04:40 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T01_57_38Z.000001.log
-rw-r----- 1 root root 104857166 Aug  2 07:35 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T04_40_01Z.000001.log
-rw-r----- 1 root root 104857144 Aug  2 10:28 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T07_35_50Z.000001.log
-rw-r----- 1 root root 104857474 Aug  2 13:16 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T10_28_18Z.000001.log
-rw-r----- 1 root root 104857529 Aug  2 16:02 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T13_16_57Z.000001.log
-rw-r----- 1 root root 104857146 Aug  2 18:52 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T16_02_15Z.000001.log
-rw-r----- 1 root root 104857342 Aug  2 21:42 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T18_52_32Z.000001.log
-rw-r----- 1 root root 104857368 Aug  3 00:26 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-02T21_42_24Z.000001.log
-rw-r----- 1 root root 104857172 Aug  3 03:23 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T00_26_06Z.000001.log
-rw-r----- 1 root root 104857158 Aug  3 06:20 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T03_23_06Z.000001.log
-rw-r----- 1 root root 104857423 Aug  3 09:13 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T06_20_37Z.000001.log
-rw-r----- 1 root root 104857547 Aug  3 12:04 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T09_13_36Z.000001.log
-rw-r----- 1 root root 104857544 Aug  3 13:57 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T12_04_58Z.000001.log
-rw-r----- 1 root root 104857563 Aug  3 15:38 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T13_57_07Z.000001.log
-rw-r----- 1 root root 104857307 Aug  3 17:28 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T15_38_40Z.000001.log
-rw-r----- 1 root root 104857344 Aug  3 19:14 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T17_28_35Z.000001.log
-rw-r----- 1 root root 104857155 Aug  3 21:00 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T19_14_40Z.000001.log
-rw-r----- 1 root root 104857472 Aug  3 22:49 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T21_00_18Z.000001.log
-rw-r----- 1 root root 104857382 Aug  4 00:36 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-03T22_49_09Z.000001.log
-rw-r----- 1 root root 104857160 Aug  4 02:19 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T00_36_59Z.000001.log
-rw-r----- 1 root root 104857581 Aug  4 04:07 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T02_19_42Z.000001.log
-rw-r----- 1 root root 104857583 Aug  4 05:52 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T04_07_27Z.000001.log
-rw-r----- 1 root root 104857357 Aug  4 07:37 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T05_52_42Z.000001.log
-rw-r----- 1 root root 104857233 Aug  4 09:25 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T07_37_46Z.000001.log
-rw-r----- 1 root root 104857302 Aug  4 11:13 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T09_25_59Z.000001.log
-rw-r----- 1 root root 104857438 Aug  4 12:55 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T11_13_57Z.000001.log
-rw-r----- 1 root root 104857537 Aug  4 14:42 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T12_55_41Z.000001.log
-rw-r----- 1 root root 104857085 Aug  4 16:24 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T14_42_44Z.000001.log
-rw-r----- 1 root root 104857199 Aug  4 18:09 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T16_24_41Z.000001.log
-rw-r----- 1 root root   7692044 Aug  4 18:16 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-04T18_09_07Z.000001.log
-rw-r----- 1 root root   6457958 Aug  5 03:43 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T03_34_50Z.000001.log
-rw-r----- 1 root root 104857440 Aug  5 06:46 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T03_43_53Z.000001.log
-rw-r----- 1 root root 104857158 Aug  5 09:27 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T06_46_06Z.000001.log
-rw-r----- 1 root root 104857395 Aug  5 12:13 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T09_27_28Z.000001.log
-rw-r----- 1 root root  81837206 Aug  5 14:28 cockroach-sql-audit.cockroachdb-db-0.root.2021-08-05T12_13_36Z.000001.log
lrwxrwxrwx 1 root root        94 Aug  5 03:43 cockroach-stderr.log -> cockroach-stderr.cockroachdb-db-0.root.2021-08-05T03_43_51Z.000001.log
-rw-r----- 1 root root      1185 Jul 26 02:49 cockroach-stderr.cockroachdb-db-0.root.2021-07-26T02_49_44Z.000001.log
-rw-r----- 1 root root      1185 Jul 26 02:53 cockroach-stderr.cockroachdb-db-0.root.2021-07-26T02_53_09Z.000001.log
-rw-r----- 1 root root      1324 Aug  5 03:43 cockroach-stderr.cockroachdb-db-0.root.2021-08-05T03_43_51Z.000001.log
drwxr-x--- 2 root root      4096 Apr 15 15:26 goroutine_dump
drwxr-x--- 2 root root     61440 Aug  5 14:15 heap_profiler
```

**Comment by knz** [comment]
ok I was able to confirm this.

@rauchenstein can you pick this up? The `gcDaemon` is not running properly for file sinks. This is a regression in v21.1.

**Comment by knz** [comment]
fixed by #68553
